src/maxflow.rs

```diff
@@ -1 +1,251 @@
+use crate::internal_queue::SimpleQueue;
+use std::cmp::min;
+use std::iter;
+use std::ops::{Add, AddAssign, Sub, SubAssign};
 
+// Maybe it should be in a separate module
+pub trait MfCapacity:
+    Copy + Ord + Add<Output = Self> + AddAssign + Sub<Output = Self> + SubAssign
+{
+    fn zero() -> Self;
+    fn max_value() -> Self;
+}
+
+impl MfCapacity for i32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i32::MAX
+    }
+}
+
+impl MfCapacity for i64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::i64::MAX
+    }
+}
+
+impl MfCapacity for u32 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u32::MAX
+    }
+}
+
+impl MfCapacity for u64 {
+    fn zero() -> Self {
+        0
+    }
+    fn max_value() -> Self {
+        std::u64::MAX
+    }
+}
+
+impl<Cap> MfGraph<Cap>
+where
+    Cap: MfCapacity,
+{
+    pub fn new(n: usize) -> MfGraph<Cap> {
+        MfGraph {
+            _n: n,
+            pos: Vec::new(),
+            g: iter::repeat_with(Vec::new).take(n).collect(),
+        }
+    }
+
+    pub fn add_edge(&mut self, from: usize, to: usize, cap: Cap) -> usize {
+        assert!(from < self._n);
+        assert!(to < self._n);
+        assert!(Cap::zero() < cap);
```

Comment:
```suggestion
        assert!(Cap::zero() < cap);
        assert_ne!(from, to);
```
or fix `rev` calculation please.

Reply:
Could you teach me why do we need this?  The original documentation says that we can input self-loops to graphs, and I don't think that the code does not work for the situation.  What change do I have to apply to `rev` calculation?

Reply:
The original implementation has the same bug-ish behavior although both this and the original impl technically "work" as these edges won't be touched. Though it is better to fix, since this bug brakes a fundamental invariant of this graph structure.

The problem is that; when `from == to`, `rev` calculated in L:68 will be the index of the edge added in `L:69`, i.e. itself instead of the reverse edge.
Since `rev` should refer the index of the reversal auxiliary edge, it has to be calculated by
```
 let rev = self.g[to].len() + if from == to { 1 } else { 0 };
```

Reply:
Ah, I've understood.  Fixing.